src/librbd/api/Group.cc

```diff
@@ -73,25 +73,145 @@ int group_snap_list(librados::IoCtx& group_ioctx, const char *group_name,
 
   const int max_read = 1024;
   cls::rbd::GroupSnapshot snap_last;
+  cls_snaps->clear();
 
   for (;;) {
     vector<cls::rbd::GroupSnapshot> snaps_page;
 
-    r = cls_client::group_snap_list(&group_ioctx, group_header_oid,
+    if (sort_support) {
+      r = cls_client::group_snap_list_sorted(&group_ioctx, group_header_oid,
+				             snap_last, max_read, &snaps_page);
+      if (r == -ENOTSUP) {
+	ldout(cct, 20) << "falling back to unsorted listing" << dendl;
```

Comment:
To expand on this, I proposed a CLS method that returns just the ordering information under the assumption that there would remain a way to produce unsorted listing for cases where sorting isn't needed (otherwise the approach not optimal and the "combined return" proposal makes more sense).  For example, `Group::remove()` that just wants to remove outstanding snapshots, or `Group::snap_remove()`, `Group::snap_rename()`, `Group::snap_rollback()` and `Group::snap_get_info()` (pending in https://github.com/ceph/ceph/pull/57797) that perform a listing a get `GroupSnapshot` struct by snapshot name.  Sorting in these cases would be redundant, so I think we need three modes of operation:

- don't try to sort (i.e. don't even call `group_snap_list_order` CLS method) -- be used in the above examples
- try to sort but don't fail if it's not possible -- to be used by `Group::snap_list()`
- try to sort and fail if it's not possible -- to be used in mirroring code

Reply:
> try to sort and fail if it's not possible -- to be used in mirroring code
By the mirroring code , do you mean the GroupReplayer or the `rbd mirror group` api?
The group replayer will need to implement the sorting with asynchronous calls so it will not call this function.

Reply:
The asynchronous helpers do not exist at the moment. Are they expected to be done in this PR ?

Reply:
No -- because there is no user in main.  But, if you want to implement it in an asynchronous fashion from the get go, you are welcome ;)  It would make integrating it into the mirroring code easier.

Reply:
I'm currently posting the PR with no asynchronous helpers and with a single additional `sort=true/false` argument for group_snap_list.

Reply:
> Just 2 modes should do - sort or not. A `rbd mirror group snap ls` (which currently does not exist) can get the sorted listing whereas everything else, including regular `group snap ls` can use sorted = no.

(Responding to a later edit of the comment.  Please try to avoid such substantial edits.)

I'm not sure what `rbd mirror group snap ls` is?  On the CLI, I would expect a single command for listing group snapshots, and we already have it -- it's `rbd group snap ls`.  It should grow `--all` argument to also include mirror snapshots, similar to how `rbd snap ls --all` behaves for images.